src/NOTD.SC2Map/MapScript.galaxy

```diff
@@ -7791,77 +7791,85 @@ void gf_CharlieCompanyAI () {
+            if ((UnitGetType(UnitGroupLoopCurrent()) == "SiegeBreaker") && (IntToFixed(UnitGroupCount(lv_siegeTanks, c_unitCountAlive)) < (IntToFixed(UnitGroupCount(gv_easyCompanyTanks, c_unitCountAlive)) * RandomFixed(0.33, 0.67))) && (lv_escortDist <= 3.0) && (lv_escortDist >= 1.35)) {
+                AISetUnitScriptControlled(UnitGroupLoopCurrent(), true);
+                UnitGroupAdd(lv_siegeTanks, UnitGroupLoopCurrent());
+                UnitIssueOrder(UnitGroupLoopCurrent(), Order(AbilityCommand("SiegeBreakerSiege", 0)), c_orderQueueReplace);
+            }
+            else if ((UnitGetType(UnitGroupLoopCurrent()) == "SiegeBreakerSieged") && ((lv_escortDist > 3.0) || (lv_escortDist < 1.35))) {
+                AISetUnitScriptControlled(UnitGroupLoopCurrent(), true);
+                UnitGroupRemove(lv_siegeTanks, UnitGroupLoopCurrent());
+                UnitIssueOrder(UnitGroupLoopCurrent(), Order(AbilityCommand("SiegeBreakerUnsiege", 0)), c_orderQueueReplace);
+                UnitGroupAdd(lv_unescorted, UnitGroupLoopCurrent());
             }
-
         }
-        lv_localIntA += 1;
+        UnitGroupLoopEnd();
+        Wait(2.1, c_timeGame);
+        UnitGroupLoopBegin(gv_easyCompany);
+        for ( ; !UnitGroupLoopDone() ; UnitGroupLoopStep() ) {
+            lv_escortDist = DistanceBetweenPoints(UnitGetPosition(UnitGroupLoopCurrent()), UnitGetPosition(UnitGroupClosestToPoint(lv_possibleEscorts, UnitGetPosition(UnitGroupLoopCurrent()))));
+            if ((lv_escortDist > 15.0)) {
+                if ((DataTableValueExists(false, ("esc_" + IntToString(UnitGetTag(UnitGroupLoopCurrent())))) == true)) {
+                    DataTableValueRemove(false, ("esc_" + IntToString(UnitGetTag(UnitGroupLoopCurrent()))));
+                }
+                AISetUnitScriptControlled(UnitGroupLoopCurrent(), true);
+                UnitIssueOrder(UnitGroupLoopCurrent(), Order(AbilityCommand("Sprint", 0)), c_orderQueueReplace);
+                UnitIssueOrder(UnitGroupLoopCurrent(), Order(AbilityCommand("Turbo", 0)), c_orderQueueReplace);
+                UnitIssueOrder(UnitGroupLoopCurrent(), OrderTargetingPoint(AbilityCommand("attack", 0), UnitGetPosition(UnitGroupClosestToPoint(lv_possibleEscorts, UnitGetPosition(UnitGroupLoopCurrent())))), c_orderQueueReplace);
+            }
+            else if ((((lv_escortDist <= 15.0) && (DataTableValueExists(false, ("esc_" + IntToString(UnitGetTag(UnitGroupLoopCurrent())))) == false)) || ((lv_escortDist > 7.0) && (DataTableValueExists(false, ("esc_" + IntToString(UnitGetTag(UnitGroupLoopCurrent())))) == true)))) {
+                UnitGroupAdd(lv_unescorted, UnitGroupLoopCurrent());
+            }
+        }
+        UnitGroupLoopEnd();
+        Wait(0.1, c_timeGame);
+        if ((UnitGroupCount(lv_unescorted, c_unitCountAlive) != 0)) {
+            UnitGroupLoopBegin(lv_unescorted);
```

Comment:
Units in question or units without escort assignment will be added to the "unassigned" unit group.
Pick each "unassigned" unit, and check if it has a designated escort (from the data table).
If there is a living escort, then assign the escort. If there is not a living escort, then assign a new one.